### PR TITLE
SUBMARINE-462. Submarine server should support helm charts

### PR DESCRIPTION
**Pull request by JohnTing**
### What is this PR for?
Currently, there are no tools to install submarine on k8s in one command.

Helm chart is a perfect tool we can choose. In order to do this:

Create helm charts supporting deployments of tensorflow operator, pytorch operator, and submarine server, MySQL.
Create Dockerfiles used to build submarine components' docker images and create tools to handle the required version push to docker hub.
Setup proxy using kubectl to access the submarine server locally
Verify that sample job spec can be submitted to the submarine server successfully.
Verify that workbench can be accessed in a local browser.
After this, if the user has a k8s cluster. The below command will get the installation done.

helm install <submarine dir>/charts

### What type of PR is it?
Feature

### Todos
* [x] - Create helm charts supporting deployments of tensorflow operator, pytorch operator, and submarine server, MySQL.
* [x] - Create Dockerfiles used to build submarine components' docker images and create tools to handle the required version push to docker hub.
* [x] - Setup proxy using kubectl to access the submarine server locally
* [x] - Verify that sample job spec can be submitted to the submarine server successfully.
* [x] - Verify that workbench can be accessed in a local browser.
* [x] - documentation

### What is the Jira issue?
https://issues.apache.org/jira/projects/SUBMARINE/issues/SUBMARINE-462

### Screenshots (if appropriate)
image
image
image
image

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by liuxunorg** [comment]
@jiwq @tangzhankun Please help review this PR, Thanks!

File: dev-support/docker-images/submarine/build.sh [hunk 1]
```diff
@@ -35,7 +35,7 @@ submarine_dist_exists=$(find -L "${SUBMARINE_HOME}/submarine-dist/target" -name
 # Build source code if the package doesn't exist.
 if [[ -z "${submarine_dist_exists}" ]]; then
   cd "${SUBMARINE_HOME}"
-  mvn clean package -DskipTests
+  mvn clean package -DskipTests -pl '!submarine-cloud'
```
[the remark this answers is not in the document]
Comment (JohnTing) [reply dev-support/docker-images/submarine/build.sh:38]: Thanks for @jiwq comment

  When I build submarine-cloud with normal user, there will be a /submarine-cloud/bin/ submarine-operator and requires root permission to delete
I have to manually delete with root permissions to proceed to the next build

**Comment by liuxunorg** [comment]
@JohnTing  This PR, wait for me to take a look.

**Comment by liuxunorg** [comment]
@JohnTing If you complete this PR, please notify me. Thanks!

**Comment by tangzhankun** [comment]
@JohnTing could you please update based on our offline discussion?
@liuxunorg , As we discussed offline, we'll first go with no-operator helm charts. Once the operator is ready. We can easily update to use operator to deploy all things. For 0.4 release, let's go with the simplified version. Does this make sense?

**Comment by JohnTing** [comment]
@tangzhankun, Thanks for your comment
There are now 3 helm charts
submarine (server & database)
tensorflow (tensorflow-operator)
pytouch (pytouch -operator)

The current version is no submarine-operator
Am I missing something?

**Comment by liuxunorg** [comment]
> @JohnTing could you please update based on our offline discussion?
> @liuxunorg , As we discussed offline, we'll first go with no-operator helm charts. Once the operator is ready. We can easily update to use operator to deploy all things. For 0.4 release, let's go with the simplified version. Does this make sense?

Good idea! i agree.

**Comment by JohnTing** [comment]
@liuxunorg 
This pr is done.
There are three helm charts.
submarine (server & database)
tensorflow (tensorflow-operator)
pytouch (pytouch-operator)
Wrote the document in setup-kubernetes.md.

**Comment by jiwq** [comment]
> tensorflow (tensorflow-operator)
> pytouch (pytouch-operator)

@JohnTing  I think we should rename the `tensorflow` to `tfjob` and `pytorch` to `pytorchjob`, because we need to be consistent with with CRD's singular name.

**Comment by tangzhankun** [comment]
@JohnTing I'm kind of confusing. Why are we making 3 helm charts? Any technical reason for that?
I was expecting a single command like "helm install submarine". Anything I missed?

**Comment by JohnTing** [comment]
@jiwq @liuxunorg I have updated this pr, please review it. Thanks.

**Comment by JohnTing** [comment]
> @JohnTing I'm kind of confusing. Why are we making 3 helm charts? Any technical reason for that?
> I was expecting a single command like "helm install submarine". Anything I missed?

A few days ago, jiwq suggested that I should divide charts into three parts https://github.com/apache/submarine/pull/270#pullrequestreview-407925752

**Comment by tangzhankun** [comment]
@JohnTing , I've discussed with the intention of multiple helm charts with Wanqiang. Only one big concern is that it seems we'd have to release multiple helms chars then?

**Comment by JohnTing** [comment]
> @JohnTing , I've discussed with the intention of multiple helm charts with Wanqiang. Only one big concern is that it seems we'd have to release multiple helms chars if we would like "helm install stable/submarine"?

No, after I update dependencies , I can merge 3 helm charts into one now.

**Comment by jiwq** [comment]
> > @JohnTing , I've discussed with the intention of multiple helm charts with Wanqiang. Only one big concern is that it seems we'd have to release multiple helms chars if we would like "helm install stable/submarine"?
> 
> No, after I update dependencies , I can merge 3 helm charts into one now.

@JohnTing Thanks for your patience and careful work. If we decide merge charts, maybe we should keep only the `submarine` charts under `helm-charts` folder. Due to charts that use `file://` are not allowed in the official Helm repository, so if we want to release it to official repo, you should merge to one chart rather use the dependencies feature.

By the way, the purpose of the split proposal is for easy maintenance and scale at later, also the submarine user can select which machine learning framework can be deployed within him/her cluster. And in the initial stage, I did not consider to release the charts to official repo.

I suggest use the split proposal, but if we want to release it to official repo, I think the single chart is the best practice.

cc: @tangzhankun

**Comment by JohnTing** [comment]
Thanks @jiwq for your suggestion.

I changed the folder structure.
Placing charts in the charts directory will also automatically deploy, 
so there is no need to set the dependency.

This is the current folder structure.
image

**Comment by tangzhankun** [comment]
@JohnTing Thanks for the hard work! @jiwq @liuxunorg Thanks for the review. Let's first go with the one helm charts without dependencies. This PR looks good to me.
Only one question, how does this image come from? Do we already have a Dockerfile for this? If so, we need to document how to build it in the document.
"apache/submarine:server-0.4.0-SNAPSHOT"

**Comment by JohnTing** [comment]
@tangzhankun 
Yes, we already have several Dockerfiles to build server and database.
I have updated setup-kubernetes.md.

**Comment by tangzhankun** [comment]
@JohnTing , Thanks for the clarification! LGTM. +1

**Comment by liuxunorg** [comment]
Will merge if no more comments.

**Pull request closed without merging**